README
WWW-Freshmeat

This modules aims at automating searches for Free Software project entries in
the Freshmeat.net directory. It is currently in very alpha stage, with very
limited functionality (retrieving data for a project you already know of).

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


Alternatively, to install with Module::Build, you can use the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc WWW::Freshmeat

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/WWW-Freshmeat

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Freshmeat

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/WWW-Freshmeat

    CPAN Ratings:
        http://cpanratings.perl.org/d/WWW-Freshmeat

COPYRIGHT AND LICENCE

Copyright 2006 Cedric Bouvier (version 0.01).
Copyright (C) 2009 Alexandr Ciornii.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for WWW-Freshmeat

0.22    2012-02-28
        - 'Mirror site' will map to 'url_mirror'
        - Compatibility support for old urls that now redirect to freecode.com

0.21    2009-12-29
        - redir_url will return string, not an URI object

0.20    2009-11-08
        - Updates for new API
        - 'retrieve_user' temporarily removed
        - "url_list", "branches", "trove_id", "real_author", "maintainers", "url", "init_html",
          "popularity" were removed
        - 'release_date' temporarily removed
        - 'languages','tags' methods added
        - example removed

0.13    not released
        - method version will return empty string when there is no release
        - new project methods: release_date, maintainers, date_add, date_updated
        - WWW::Freshmeat::User added
        - example added

0.12    2009-02-18
        - added $WWW::Freshmeat::Project::project_re - regexp for project name
        - url_list will work with projects containing '.'
        - added real_author to fetch name of author (not maintainer)

0.11    2009-02-15
        - url_list has argument to try to follow Freshmeat redirect links
        - rewrote redir_url not to fetch page it was redirected to

0.10    2009-02-14
        - 5.008 only (due to 'ISO-8859-1' encoding of XML file)
        - projectname_short project method for retrieving project string id
        - HTML::TreeBuilder::XPath is now required
        - supports fetching branch list
        - url_list project method for retrieving project URL list
        - popularity project method for retrieving project popularity data
        - redir_url in WWW::Freshmeat to fetch url which link will redirect to

0.03    2009-02-03
        - retrieve_project for project that does not exist will return undef
        - retrieve_project will die if it would not be able to fetch project record
        - nonexistingproject.t to test this functionality
        - method for returning id's of trove categories
        - project_from_xml class method
        - tests check for connection with LWP::Online
        - Build.PL now lists modules in 'requires' section

0.02    2009-01-25
        new maintainer: Alexandr Ciornii
        - distribution modernized
        - "project.t" added, a live test
        - "boilerplate.t" removed
        - "all rights reserved" removed (by permission of Cedric Bouvier)

0.01    Wed May 31 13:29:17 CEST 2006
        First version, only retrieval of known project entry.



Makefile.PL
use 5.008;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'WWW::Freshmeat',
    AUTHOR              => 'Alexandr Ciornii <alexchorny@gmail.com>',
    VERSION_FROM        => 'lib/WWW/Freshmeat.pm',
    ABSTRACT_FROM       => 'lib/WWW/Freshmeat.pm',
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' => 0.17, #isa_ok
        'LWP::UserAgent' => 0,
        'XML::Simple' => 0,
        'XML::Parser' => 0, #for XML::Simple, in case of dependency problems
        'LWP::Online' => 1.06,
        #'HTML::TreeBuilder::XPath' => 0.09,
        'Mouse' => 0.24, #recent one
        'File::Slurp' => 0,
    },
    ($ExtUtils::MakeMaker::VERSION ge '6.31'? 
     ('LICENSE'		=> 'perl', ) : ()),
    ($ExtUtils::MakeMaker::VERSION ge '6.48'? 
     ('MIN_PERL_VERSION' => 5.008,) : ()),
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'WWW-Freshmeat-*' },
);


lib/WWW/Freshmeat.pm
package WWW::Freshmeat;

use 5.008;
use strict;
use warnings;

=head1 NAME

WWW::Freshmeat - automates usage of Freshmeat.net

=head1 VERSION

Version 0.22

=cut

our $VERSION = '0.22';

use XML::Simple qw();
use WWW::Freshmeat::Project;
use Carp;


=head1 SYNOPSIS

    use WWW::Freshmeat;

    my $fm = WWW::Freshmeat->new(token=>'freshmeat_token');

    my $project = $fm->retrieve_project('project_id');

    foreach my $p ( @projects, $project ) {
        print $p->name(), "\n";
        print $p->version(), "\n";
        print $p->description(), "\n";
    }

=cut

package WWW::Freshmeat;

use base qw( LWP::UserAgent );

sub new {
  my $class=shift;
  my $self=LWP::UserAgent->new();
  bless $self,$class;
  my %data=@_;
  $self->{fm_token}=$data{token};
  return $self;
}

sub _token {
  my $self = shift;
  croak "No token" unless $self->{fm_token};
  return $self->{fm_token};
}

=head1 DESCRIPTION

C<WWW::Freshmeat> derives from C<LWP::UserAgent>, so it accepts all the methods
that C<LWP::UserAgent> does, notably C<timeout>, C<useragent>, C<env_proxy>...

=head2 Methods

=over 4

=item B<retrieve_project> I<STRING>

Query the freshmeat.net site for the project I<STRING> (should be the Freshmeat
ID of the requested project) and returns a C<WWW::Freshmeat::Project> object or
undef if the project entry cannot be found.

=cut

sub retrieve_project {
    my $self = shift;
    my $id   = shift;

    my $url = "http://freshmeat.net/projects/$id.xml?auth_code=".$self->_token;

    my $response = $self->get($url);
    if ($response->is_success) {
      my $xml = $response->content();
      return $self->project_from_xml($xml);
    } else {
      if ($response->code eq '404') {
        return undef;
      } else {
        die "Could not GET freshmeat project (".$response->status_line.")";
      }
    }
}

=item B<project_from_xml> I<STRING>

Receives Freshmeat project XML record and returns a C<WWW::Freshmeat::Project>
object or undef if the project entry cannot be found.

=cut

sub project_from_xml {
    my $self = shift;
    my $xml  = shift;

    if ($xml eq 'Error: project not found.') {
      return undef;
    }
    die "XML is empty" unless $xml;

    my $data = XML::Simple::XMLin($xml,ForceArray => ['approved-url','recent-release']);
    #die unless exists $data->{'project'};
    die unless $data->{'name'};

    return WWW::Freshmeat::Project->new($data, $self); #->{'project'}
}

sub retrieve_user {
    croak "'User' is temporarily removed";
    my $self = shift;
    my $id   = shift;
    require WWW::Freshmeat::User;
    return WWW::Freshmeat::User->new($self,$id);
}

=item B<redir_url> I<STRING>

Receives URL and returns URL which it redirects to.

=cut

sub redir_url {
    my $self = shift;
    my $url=shift;
    $self->requests_redirectable([]);
    my $response = $self->get($url) or return $url;
    if ($response->is_redirect) {
      #http://www.perlmonks.org/?node_id=147608
      my $referral_uri = $response->header('Location');
      {
          # Some servers erroneously return a relative URL for redirects,
          # so make it absolute if it not already is.
          local $URI::ABS_ALLOW_RELATIVE_SCHEME = 1;
          my $base = $response->base;
          $referral_uri = $HTTP::URI_CLASS->new($referral_uri, $base)
                      ->abs($base)->as_string;
      }
      return $referral_uri;
    } else {
      return $url;
    }
}

=back

=head1 SEE ALSO

L<LWP::UserAgent>.

=head1 AUTHOR

Cedric Bouvier, C<< <cbouvi at cpan.org> >>. Alexandr Ciornii.

=head1 BUGS

Please report any bugs or feature requests to
C<bug-www-freshmeat at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Freshmeat>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Freshmeat

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Freshmeat>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Freshmeat>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Freshmeat>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Freshmeat>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2006 Cedric Bouvier (version 0.01).
Copyright 2009-2012 Alexandr Ciornii.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of WWW::Freshmeat


examples/popularity.pl
#!/usr/bin/perl

use strict;
use warnings;

use WWW::Freshmeat 0.20;
die "Will not work with current version of WWW::Freshmeat\n";
=for cmt
  use Getopt::Long;
  my $user;
  my $date;
  my $fp;
  my $result = GetOptions ("user=s" => \$user,
                           "added_after=s" => \$date, #only projects added after this date
                           "first-page=s" => \ $fp,
                           #Number on front page, name\tvalue\n
  );
die "Please specify user\n" unless $user;

my $fm = WWW::Freshmeat->new;

my $user_p = $fm->retrieve_user('chorny');

my @projects=$user_p->projects();
my @proj=map {$fm->retrieve_project($_)} @projects;
my %fp_num;
if ($fp) {
  open my $fp_in,'<',$fp or die "Cannot open $fp\n";
  while (my $str=<$fp_in>) {
    $str=~s/\s+$//s;
    next unless $str;
    if ($str=~/^\s*([^\t]+)\t([^\t\r\n]+)$/) {
      $fp_num{$1}=$2;
    } else {
      die "Bad string in $fp: '$str'";
    }
  }
}

open my $out,'>','popularity.html';
print $out <<EOT;
<title>Freashmeat popularity data for user '$user'</title>
<style>
table.fm_stat tr td:first-child + td + td,
table.fm_stat tr td:first-child + td + td + td ,
table.fm_stat tr td:first-child + td + td + td + td,
table.fm_stat tr td:first-child + td + td + td + td + td
{
 text-align: right;
}
</style>

<table border=1 class="fm_stat">
 <tr>
  <th>Name</th>
  <th>Creation date</th>
  <th>Record hits</th>
  <th>URL hits</th>
  <th>Subscribers</th>
  <th>Number on front page</th>
 </tr>
EOT

my $total_record_hits=0;
my $total_url_hits=0;
my $total_subscribers=0;
foreach my $project (@proj) {
  if ($date) {
    next if $date ge $project->date_add();
  }
  my %pop=$project->popularity();
  my $fp_p=
   exists $fp_num{$project->name()}?
     $fp_num{$project->name()}:
   exists $fp_num{$project->projectname_short()}?
     $fp_num{$project->projectname_short()}:
   '?';
  $total_record_hits+=$pop{'record_hits'};
  $total_url_hits+=$pop{'url_hits'};
  $total_subscribers+=$pop{'subscribers'};
  print $out "
 <tr>
   <td><a href='http://freshmeat.net/project-stats/view/".$project->{project_id}."/'>".
    $project->name()."</a></td>
   <td>".$project->date_add()."</td>
   <td>$pop{'record_hits'}</td>
   <td>$pop{'url_hits'}</td>
   <td>$pop{'subscribers'}</td>
   <td>$fp_p</td>
 </tr>
\n";
}
  print $out "
 <tr>
   <td>Total</td>
   <td>-</td>
   <td>$total_record_hits</td>
   <td>$total_url_hits</td>
   <td>$total_subscribers</td>
   <td>-</td>
 </tr>
\n";

print $out <<'EOT';
</table>
EOT
=cut

lib/WWW/Freshmeat/User.pm
package WWW::Freshmeat::User;

use strict;
use warnings;
use WWW::Freshmeat ();
use Carp;

our $VERSION = '0.22';

sub init_html {
    croak "removed";
    my $self = shift;
    my $html = shift;
    require HTML::TreeBuilder::XPath;
    $self->{_html}=HTML::TreeBuilder::XPath->new_from_content($html);
}

sub _html_tree {
    my $self = shift;
    if (!$self->{_html}) {
      my $id=$self->{id};
      my $url = "http://freshmeat.net/~$id/";
      $self->{www_freshmeat}->agent('User-Agent: Mozilla/5.0 (Windows; U; Windows NT 5.1; ru; rv:1.8.1.19) Gecko/20081201 Firefox/2.0.0.19');
      my $response = $self->{www_freshmeat}->get($url);
      my $html = $response->content();
      if ($response->is_success) {
        if ($html=~m#<p>\s*Invalid user ID\s*</p>#s) {
          return;
        }
        $self->init_html($html);
      } else {
        die "Could not GET $url (".$response->status_line.", $html)";
      }
    }
    return $self->{_html};
}

sub new {
    croak "removed";
    my $proto = shift;
    my $class = ref($proto) || $proto;

    my $self = bless {}, $class;
    $self->{www_freshmeat} = shift;
    $self->{id} = shift;
    return $self;
}


sub projects {
    my $self = shift;
    my $tree=$self->_html_tree();
    my $xpath=q{/html/body/div[1]/table/tr/td[2]/table/tr[1]/td[2]/p/ul[1]/li/a};
    my $nodes=$tree->findnodes($xpath);
    my @list;
    while (my $node=$nodes->shift) {
      if ($node->attr('href') =~m#/projects/($WWW::Freshmeat::Project::project_re)/#) {
        push @list,$1;
      } else {
        die;
      }
    }
    return @list;
}

1

# (c) Alexandr Ciornii, 2009


Build.PL
use 5.008;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'WWW::Freshmeat',
    license             => 'perl',
    dist_author         => 'Alexandr Ciornii <alexchorny@gmail.com>',
    dist_version_from   => 'lib/WWW/Freshmeat.pm',
    requires => {
        'LWP::UserAgent' => 0,
        'XML::Simple' => 0,
        'XML::Parser' => 0, #for XML::Simple, in case of dependency problems
        'perl' => 5.008,
        #'HTML::TreeBuilder::XPath' => 0.09,
        'Mouse' => 0.24, #recent one
    },
    build_requires => {
        'Test::More' => 0.17, #isa_ok
        'LWP::Online' => 1.06,
        'File::Slurp' => 0,
    },
    meta_merge => {
        resources=> {
            repository => 'http://github.com/chorny/WWW-Freshmeat',
            Ohloh => 'https://www.ohloh.net/p/WWW-Freshmeat',
        },
        keywords => ['freshmeat','API'],
    },
    add_to_cleanup      => [ 'WWW-Freshmeat-*' ],
);

$builder->create_build_script();
